ethereum/extethclient/eth_wallet_client.go

```diff
@@ -148,37 +149,56 @@ func (c *ethClient) SuggestGasPrice(ctx context.Context) (*big.Int, error) {
 	return c.Raw().SuggestGasPrice(ctx)
 }
 
-func (c *ethClient) ERC20Balance(ctx context.Context, token ethcommon.Address) (*big.Int, error) {
-	tokenContract, err := contracts.NewIERC20(token, c.ec)
+func (c *ethClient) ERC20Balance(ctx context.Context, tokenAddr ethcommon.Address) (*coins.ERC20TokenAmount, error) {
+	tokenContract, err := contracts.NewIERC20(tokenAddr, c.ec)
 	if err != nil {
-		return big.NewInt(0), err
+		return nil, err
+	}
+
+	bal, err := tokenContract.BalanceOf(c.CallOpts(ctx), c.Address())
+	if err != nil {
+		return nil, err
 	}
-	return tokenContract.BalanceOf(c.CallOpts(ctx), c.Address())
+
+	tokenInfo, err := c.erc20Info(ctx, tokenAddr, tokenContract)
+	if err != nil {
+		return nil, err
+	}
+
+	return coins.NewERC20TokenAmountFromBigInt(bal, tokenInfo), nil
 }
 
-func (c *ethClient) ERC20Info(ctx context.Context, token ethcommon.Address) (
-	name string,
-	symbol string,
-	decimals uint8,
-	err error,
-) {
-	tokenContract, err := contracts.NewIERC20(token, c.ec)
+func (c *ethClient) erc20Info(
+	ctx context.Context,
+	tokenAddr ethcommon.Address,
+	tokenContract *contracts.IERC20,
+) (*coins.ERC20TokenInfo, error) {
+	name, err := tokenContract.Name(c.CallOpts(ctx))
 	if err != nil {
-		return "", "", 18, err
+		return nil, err
 	}
-	name, err = tokenContract.Name(c.CallOpts(ctx))
+
+	symbol, err := tokenContract.Symbol(c.CallOpts(ctx))
 	if err != nil {
-		return "", "", 18, err
+		return nil, err
 	}
-	symbol, err = tokenContract.Symbol(c.CallOpts(ctx))
+
+	// TODO: Do we support ERC20 tokens that do not have this method?
```

Comment:
are decimals 18 if they're not specified? not sure though

Reply:
With the normal ERC20 contracts, the decimals function will always be present and it will default to 18 decimals if you don't configure it otherwise in the constructor. As best as I can guess from [the standard](https://eips.ethereum.org/EIPS/eip-20), if the decimals function does not exist on the contract, you would treat it as zero decimals and there would never be a fractional part on the user representation. (But the standard doesn't explicitly state that, and I was unable to find any deployed token where the function is missing.)